### PR TITLE
Distributed pm

### PR DESCRIPTION
**Pull request by ASKabalan**
(Draft)

On going work to make it dead simple to do distributed PM simulations

### CODE REVIEW

**Comment by ASKabalan** [comment]
@EiffL 

This my ongoing work on making a simple distribution-agnostic code for PM simulations
Most of the magic happens in https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/distributed_pm/jaxpm/_src/spmd_config.py

**Comment by ASKabalan** [comment]
@EiffL @aboucaud 
Still some work to do but the exemple is 'almost' working.
I think I rather use what François suggested to create distributed arrays
(Using a on the fly created shardmap rather than a callback)
This why, we can close on arrays and still have them be lowered to ir.RankedTensor
The CallBackOperator is the one still causing issues

**Comment by EiffL** [comment]
ok, so, @ASKabalan, you did a lot of work here, I appreciate it. But your proposal adds a lot of boiller code that is not necessary. 

My hope for jaxpm is to have code that is as close to plain jax as possible, and as little infrastructure code as possible. All what the library should provide is what is needed to define the ode function, and we let the user free to use whatever ode solver they want. Essentially, making JaxPM the leanest and simplest library in the west.  

I prototyped an alternative proposal that makes the smallest amount of modifications to the existing code, and which (I think) does not make significant changes to the existing API. 
Check it out here: https://github.com/DifferentiableUniverseInitiative/JaxPM/compare/jaxdecomp_proto?expand=1 
This only requires like 10 lines of code in addition to a couple of utility functions.

Then all what people would need to do is to execute the code inside a `with mesh` environment:

```python
mesh_shape = [256, 256, 256]
box_size = [256., 256., 256.]
snapshots = jnp.linspace(0.1, 1., 2)

@jax.jit
def run_simulation(omega_c, sigma8, seed):
    # Create a cosmology
    cosmo = jc.Planck15(Omega_c=omega_c, sigma8=sigma8)

    # Create a small function to generate the matter power spectrum
    k = jnp.logspace(-4, 1, 128)
    pk = jc.power.linear_matter_power(
        jc.Planck15(Omega_c=omega_c, sigma8=sigma8), k)
    pk_fn = lambda x: jc.scipy.interpolate.interp(x.reshape([-1]), k, pk
                                                  ).reshape(x.shape)

    # Create initial conditions
    initial_conditions = linear_field(mesh_shape, box_size, pk_fn, seed=seed)

    # Initialize particle displacements
    dx, p, f = lpt(cosmo, initial_conditions, 1.0)

    field = cic_paint(jnp.zeros_like(initial_conditions), dx)
    return field


def main(args):
    # Setting up distributed random numbers
    master_key = jax.random.PRNGKey(42)
    key = jax.random.split(master_key, size)[rank]

    # Create computing mesh and sharding information
    devices = mesh_utils.create_device_mesh((2, 2))
    mesh = Mesh(devices.T, axis_names=('x', 'y'))

    # Run the simulation on the compute mesh
    with mesh:
        field = run_simulation(0.32, 0.8, key)

    print('done')
    np.save(f'field_{rank}.npy', field.addressable_data(0))

    # Closing distributed jax
    jax.distributed.shutdown()
```

**Comment by EiffL** [comment]
The example in that branch is not complete, but it's just to illustrate the approach.

**Comment by ASKabalan** [comment]
Ok I understand your approach.
I was fixated on solving the named axis problem.
I guess if we impose the axis name I would have to do much less code.
I can take it from here I see what needs to be done
You confirm that we will impose the axis names?

**Comment by EiffL** [comment]
Yes :-) confirmed. Axis names, 'x', 'y', 'z' are conventions of the library.

**Comment by ASKabalan** [comment]
closing this PR
